### PR TITLE
fix: popups same width as parent

### LINKED ISSUE #6609: [Enhancement] Select widget and other popups should be the size of the parent
## Summary

The popup for a widget should be the size of the parent until the parent width is above the min width of the popup

### PR DESCRIPTION
**Pull request by Tooluloope**
## Description

Make popups have a min-width as the Parent Container (Select and Menubutton WIdget)

Fixes #6609 

## Type of change

> Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: fix/popups-same-width-as-parent 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.07 **(-0.02)** | 37.01 **(-0.01)** | 33.85 **(-0.01)** | 55.56 **(-0.02)**
 :green_circle: | app/client/src/widgets/WidgetUtils.ts | 84.62 **(0.31)** | 59.38 **(0)** | 54.55 **(0)** | 80.95 **(0.46)**
 :green_circle: | app/client/src/widgets/DropdownWidget/component/index.tsx | 82.67 **(1.51)** | 50.72 **(-0.05)** | 70 **(1.58)** | 81.43 **(0.83)**
 :green_circle: | app/client/src/widgets/DropdownWidget/widget/index.tsx | 72.92 **(1.18)** | 23.53 **(0)** | 63.64 **(0)** | 72.09 **(1.36)**
 :red_circle: | app/client/src/widgets/MenuButtonWidget/component/index.tsx | 25 **(-0.93)** | 13.89 **(-0.82)** | 0 **(0)** | 30.61 **(0.18)**
 :red_circle: | app/client/src/widgets/MenuButtonWidget/widget/index.tsx | 58.62 **(-2.92)** | 0 **(0)** | 42.86 **(0)** | 59.26 **(-3.24)**
 :red_circle: | app/client/src/widgets/MultiSelectTreeWidget/component/index.styled.tsx | 45.24 **(-4.76)** | 24.24 **(-0.76)** | 0 **(0)** | 50 **(-1.52)**
 :green_circle: | app/client/src/widgets/MultiSelectTreeWidget/component/index.tsx | 25.49 **(1.58)** | 0 **(0)** | 0 **(0)** | 27.66 **(1.47)**
 :red_circle: | app/client/src/widgets/MultiSelectTreeWidget/widget/index.tsx | 43.55 **(-1.45)** | 0 **(0)** | 42.86 **(0)** | 44.83 **(-1.6)**
 :red_circle: | app/client/src/widgets/MultiSelectWidget/component/index.styled.tsx | 42.86 **(-8.86)** | 23.33 **(-0.81)** | 0 **(0)** | 48.28 **(-3.57)**
 :green_circle: | app/client/src/widgets/MultiSelectWidget/component/index.tsx | 19.7 **(1.37)** | 0 **(0)** | 0 **(0)** | 20.97 **(1.33)**
 :red_circle: | app/client/src/widgets/MultiSelectWidget/widget/index.tsx | 48.08 **(-1.92)** | 0 **(0)** | 50 **(0)** | 48.98 **(-2.08)**
 :red_circle: | app/client/src/widgets/SingleSelectTreeWidget/component/index.styled.tsx | 47.5 **(-5.28)** | 25 **(-0.81)** | 0 **(0)** | 51.52 **(-1.61)**
 :green_circle: | app/client/src/widgets/SingleSelectTreeWidget/component/index.tsx | 26 **(1.56)** | 0 **(0)** | 0 **(0)** | 28.26 **(1.43)**
 :red_circle: | app/client/src/widgets/SingleSelectTreeWidget/widget/index.tsx | 51.85 **(-2)** | 0 **(0)** | 50 **(0)** | 54 **(-2.25)**</details>

### CODE REVIEW

File: app/client/src/widgets/DropdownWidget/component/index.tsx [hunk 1]
```diff
@@ -153,7 +154,7 @@ class DropDownComponent extends React.Component<DropDownComponentProps> {
   render() {
     return (
       <DropdownContainer>
-        <DropdownStyles />
+        <DropdownStyles parentWidth={this.props.width - 7} />
```
Comment (somangshu) [inline app/client/src/widgets/DropdownWidget/component/index.tsx:157]: Why do we subtract a hardcoded value here?

File: app/client/src/widgets/MenuButtonWidget/component/index.tsx [hunk 2]
```diff
@@ -114,10 +114,14 @@ export const MenuButtonContainer = styled.div`
...
 `;
 
-const PopoverStyles = createGlobalStyle`
+const PopoverStyles = createGlobalStyle<{ parentWidth: number }>`
   .menu-button-popover > .${Classes.POPOVER2_CONTENT} {
     background: none;
   }
+  .menu-button-popover {
+     min-width:${({ parentWidth }) =>
+       parentWidth > 180 ? `${parentWidth}px` : "180px"} ;
```
Comment (somangshu) [inline app/client/src/widgets/MenuButtonWidget/component/index.tsx:123]: For the sake of consistency, we should have min width consistent for all the widget popups

@jsartisan @momcilo-appsmith need your help reviewing this.

File: app/client/src/widgets/DropdownWidget/component/index.tsx [hunk 3]
```diff
@@ -88,9 +88,10 @@ const StyledControlGroup = styled(ControlGroup)<{ haslabel: string }>`
   }
 `;
 
-const DropdownStyles = createGlobalStyle`
+const DropdownStyles = createGlobalStyle<{ parentWidth: number }>`
   .select-popover-wrapper {
-    width: 100%;
+    min-width:${({ parentWidth }) =>
+      parentWidth > 269 ? `${parentWidth}px` : "269px"} ;
```
Comment (somangshu) [inline app/client/src/widgets/DropdownWidget/component/index.tsx:94]: The behaviour is still different for the multi-select widget.

@marks0351 does this look like the right way to move forward?

**Review by somangshu — DISMISSED** [review]
Approving to run tests

**Comment by somangshu** [comment]
@Tooluloope can we add cypress tests for these?

File: app/client/src/widgets/DropdownWidget/component/index.tsx (hunk 3 above)
Comment (marks0351) [reply app/client/src/widgets/DropdownWidget/component/index.tsx:94]: @Tooluloope what does the value 269 indicate?

File: app/client/src/widgets/MenuButtonWidget/component/index.tsx (hunk 2 above)
Comment (marks0351) [reply app/client/src/widgets/MenuButtonWidget/component/index.tsx:123]: @Tooluloope what does 180 indicate?

**Review by marks0351 — CHANGES_REQUESTED** [review]

File: app/client/src/widgets/MenuButtonWidget/component/index.tsx (hunk 2 above)
Comment (Tooluloope) [reply app/client/src/widgets/MenuButtonWidget/component/index.tsx:123]: Current min-width of the popups

File: app/client/src/widgets/DropdownWidget/component/index.tsx (hunk 3 above)
Comment (Tooluloope) [reply app/client/src/widgets/DropdownWidget/component/index.tsx:94]: Current min-width of the popups

Comment (Tooluloope) [reply app/client/src/widgets/DropdownWidget/component/index.tsx:94]: We could have a consistent value, but I'm not sure it would valid for all popups

**Comment by momcilo-appsmith** [comment]
@somangshu the latest deploy preview is not working for me.

**Comment by Tooluloope** [comment]
@momcilo-appsmith here the [link](https://deploy-preview-7499--appsmith-editor-ce.netlify.app/user/login)

File: app/client/src/widgets/DropdownWidget/component/index.tsx (hunk 1 above)
Comment (Tooluloope) [reply app/client/src/widgets/DropdownWidget/component/index.tsx:172]: @somangshu The space between the Component itself and the Widget Container

Screenshot 2021-09-28 at 13 32 32

**Comment by Tooluloope** [comment]
@momcilo-appsmith have you been able to use the Deploy preview link?

**Comment by Tooluloope** [comment]
@somangshu @marks0351 Here's an update on the Dropdown https://www.figma.com/file/w9CaXHF5iYcrLtiEbAHcFu/Original-Theme-Widgets-Library?node-id=428%3A3414

**Comment by Tooluloope** [comment]
@somangshu @marks0351 Here's an update on the Dropdown https://www.figma.com/file/w9CaXHF5iYcrLtiEbAHcFu/Original-Theme-Widgets-Library?node-id=428%3A3414

**Review by marks0351 — APPROVED** [review]

**Comment by vivekverma2312** [comment]
@Tooluloope 1. Minimum width of pop up for different widget is different (In screenshot I have set the widget size to there minimum width)
Screenshot 2021-11-01 at 1 45 28 PM

2. Pop up size for menu button changes after we drag and drop another Menu button widget on canvas
Screenshot 2021-11-01 at 1 47 56 PM
Screenshot 2021-11-01 at 1 47 39 PM

3. Pop up is of not the same size as widget when  placed in any container

Screenshot 2021-11-01 at 1 51 34 PM

**Comment by Tooluloope** [comment]
@vivekverma2312 Fixed ALL

**Comment by marks0351** [comment]
> @Tooluloope 1. Minimum width of pop up for different widget is different (In screenshot I have set the widget size to there minimum width) Screenshot 2021-11-01 at 1 45 28 PM
> 
> 1. Pop up size for menu button changes after we drag and drop another Menu button widget on canvas
> 
> Screenshot 2021-11-01 at 1 47 56 PM
> 
> Screenshot 2021-11-01 at 1 47 39 PM
> 
> 1. Pop up is of not the same size as widget when  placed in any container
> 
> Screenshot 2021-11-01 at 1 51 34 PM

@Tooluloope were these not tested in the test cases?

**Comment by Tooluloope** [comment]
I tested for only Select and Menu button, DIdn't know MUltiSelect, MultiTreeSelect, and TreeSelect was to be covered.

PR description states only Select and Menubutton

**Comment by marks0351** [comment]
> I tested for only Select and Menu button, DIdn't know MUltiSelect, MultiTreeSelect, and TreeSelect was to be covered.
> 
> PR description states only Select and Menubutton

got it. can we pls add tests for them as well, will make sure we have protection over these widgets as well.

**Comment by vivekverma2312** [comment]
@Tooluloope  Above issues are fixed but observed other issues now
4. Pop up size is not same for Menu Button widget inside the container by default
Screenshot 2021-11-03 at 12 10 22 PM

5. The drop down options for multi-select widget inside container is not coming proper
Screenshot 2021-11-03 at 12 09 45 PM

**Comment by vivekverma2312** [comment]
@Tooluloope Fix are working fine on Deploy Preview.

**Review by marks0351 — DISMISSED** [review]

**Review by marks0351 — DISMISSED** [review]

**Comment by marks0351** [comment]
@Tooluloope can you pls resolve conflicts

**Review by marks0351 — APPROVED** [review]

**Review by marks0351 — APPROVED** [review]

**Pull request merged**